core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1058,49 +1058,90 @@ struct Block : protected std::tuple<Arguments...> {
+                                                                               : std::min(ports_status.in_samples, ports_status.nSamplesToNextTag);
+                ports_status.out_samples = ports_status.in_samples;
+            }
+#ifdef _DEBUG
+            fmt::println("block {} - mark2 - in {} -> out {}", name, ports_status.in_samples, ports_status.out_samples);
+#endif
+            ports_status.in_samples = std::min(ports_status.in_samples, ports_status.nSamplesToEosTag); // clamp input until EOS tag
+            // TODO: recheck definitions of numerator and denominator
+            //   Provided we we define N := <numerator> and M := <denominator>, do we expect (assuming a simple in->out block) that
+            //  a) for N input samples M output samples are produced, or
+            //  b) for a given number of input samples X -> X * N/M output samples are produced.
+            //  Some of the logic conditions vary depending on whether we assume 'a)' or 'b)'.
+            const bool isEOSTagPresent = ports_status.nSamplesToEosTag == 0                                                             //
+                                      || ports_status.nSamplesToEosTag < std::max(ports_status.in_samples, ports_status.in_min_samples) //
+                                      || ports_status.nSamplesToEosTag < numerator;
+#ifdef _DEBUG
+            if (isEOSTagPresent) {
+                fmt::println("##block {} received EOS tag at {} < in_min_samples {}", name, ports_status.nSamplesToEosTag, ports_status.in_min_samples);
+            }
+#endif
+
+            if (isEOSTagPresent || lifecycle::isShuttingDown(state)) {
+                state = lifecycle::State::STOPPED;
+                state.notify_all();
+#ifdef _DEBUG
+                fmt::println("##block {} received EOS tag at {} in_samples {} -> lifecycle::State::STOPPED", name, ports_status.nSamplesToEosTag, ports_status.in_samples);
+#endif
+                updateInputAndOutputTags(static_cast<Tag::signed_index_type>(ports_status.in_min_samples));
                 updateOutputTagsWithSettingParametersIfNeeded();
-                forwardTags();
                 return { requested_work, 0UZ, work::Status::DONE };
             }
 
             if constexpr (Resampling::kEnabled) {
-                const bool isEOSTagPresent  = ports_status.in_samples_to_eos_tag < ports_status.in_samples;
-                const auto rasamplingStatus = doResampling();
-                if (rasamplingStatus != work::Status::OK) {
-                    if (rasamplingStatus == work::Status::INSUFFICIENT_INPUT_ITEMS && isEOSTagPresent) {
-                        state = LifeCycleState::STOPPED;
-                        updateInputAndOutputTags();
+                const auto resamplingStatus = doResampling();
+                if (resamplingStatus != work::Status::OK) {
+                    if (resamplingStatus == work::Status::INSUFFICIENT_INPUT_ITEMS || isEOSTagPresent) {
```

Comment:
I think there is a logic error here.
`isEOSTagPresent` is always false when reached this `if`.
if it's true `if (isEOSTagPresent || lifecycle::isShuttingDown(state))` is also true and it `return`s.

`isEOSTagPresent` should be separated for `Resampling::kEnabled`